src/qibo/transpiler/unitary_decompositions.py

```diff
@@ -0,0 +1,254 @@
+    ef = (psi_bar[0] + 1j * psi_bar[1]) / np.sqrt(2)
+    e_f_ = (psi_bar[0] - 1j * psi_bar[1]) / np.sqrt(2)
+    e, f = schmidt_decompose(ef)
+    e_, f_ = schmidt_decompose(e_f_)
+    # find exp(1j * delta) using (A5a)
+    ef_ = np.kron(e, f_)
+    phase = 1j * np.sqrt(2) * np.dot(np.conj(ef_), psi_bar[2])
+
+    # construct unitaries UA, UB using (A6a), (A6b)
+    ua = np.tensordot([1, 0], np.conj(e), axes=0) + phase * np.tensordot(
+        [0, 1], np.conj(e_), axes=0
+    )
+    ub = np.tensordot([1, 0], np.conj(f), axes=0) + np.conj(phase) * np.tensordot(
+        [0, 1], np.conj(f_), axes=0
+    )
+    return ua, ub
+
+
+def calculate_diagonal(unitary, ua, ub, va, vb):
+    """Calculates Ud matrix that can be written as exp(-iH).
+
+    See Eq. (A1) in arXiv:quant-ph/0011050.
+    Ud is diagonal in the magic and Bell basis.
+    """
+    # normalize U_A, U_B, V_A, V_B so that detU_d = 1
+    # this is required so that sum(lambdas) = 0
+    # and Ud can be written as exp(-iH)
+    det = np.linalg.det(unitary) ** (1 / 16)
+    ua *= det
+    ub *= det
+    va *= det
+    vb *= det
+    u_dagger = np.conj(np.kron(ua, ub).T)
+    v_dagger = np.conj(np.kron(va, vb).T)
+    ud = np.dot(np.dot(u_dagger, unitary), v_dagger)
+    return ua, ub, ud, va, vb
+
+
+def magic_decomposition(unitary):
+    """Decomposes an arbitrary unitary to (A1) from arXiv:quant-ph/0011050."""
+    psi, eigvals = calculate_psi(unitary)
+    psi_tilde = np.conj(np.sqrt(eigvals)) * np.dot(unitary, psi)
+    va, vb = calculate_single_qubit_unitaries(psi)
+    ua_dagger, ub_dagger = calculate_single_qubit_unitaries(psi_tilde)
+    ua, ub = np.conj(ua_dagger.T), np.conj(ub_dagger.T)
+    return calculate_diagonal(unitary, ua, ub, va, vb)
+
+
+def to_bell_diagonal(ud):
+    """Transforms a matrix to the Bell basis and checks if it is diagonal."""
+    ud_bell = np.dot(np.dot(np.conj(bell_basis).T, ud), bell_basis)
+    ud_diag = np.diag(ud_bell)
+    if not np.allclose(np.diag(ud_diag), ud_bell):  # pragma: no cover
+        return None
+    uprod = np.prod(ud_diag)
+    if not np.allclose(uprod, 1):  # pragma: no cover
+        return None
+    return ud_diag
+
+
+def calculate_h_vector(ud_diag):
+    """Finds h parameters corresponding to exp(-iH).
+
+    See Eq. (4)-(5) in arXiv:quant-ph/0307177.
+    """
+    lambdas = -np.angle(ud_diag)
+    hx = (lambdas[0] + lambdas[2]) / 2.0
+    hy = (lambdas[1] + lambdas[2]) / 2.0
+    hz = (lambdas[0] + lambdas[1]) / 2.0
+    return hx, hy, hz
+
+
+def cnot_decomposition(q0, q1, hx, hy, hz):
+    """Performs decomposition (6) from arXiv:quant-ph/0307177."""
+    u3 = -1j * (matrices.X + matrices.Z) / np.sqrt(2)
+    # use corrected version from PRA paper (not arXiv)
+    u2 = -u3 @ expm(-1j * (hx - np.pi / 4) * matrices.X)
+    # add an extra exp(-i pi / 4) global phase to get exact match
+    v2 = expm(-1j * hz * matrices.Z) * np.exp(-1j * np.pi / 4)
+    v3 = expm(1j * hy * matrices.Z)
```

Comment:
These can be defined, respectively, as $R_{X}(2 h_{x} - \\pi / 2)$, $e^{-i\\pi/4} \\, R_{Z}(2 h_{z})$, and $R_{Z}(-h_{y})$